File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const [m, n] = [matrix.length, matrix[0].length];
    function bin(row, i, j) {
        const mid = Math.floor((i + j) / 2);
        if (target === matrix[row][mid]) return true;
        if (mid === i || mid === j) return false;
        if (matrix[row][mid] < target) {
            return bin(row, mid, j);
        } else {
            return bin(row, i, mid);
        }
    }
    for (let i = 0; i < m; i++) {
        if (target < matrix[i][0]) return false;
        if (target > matrix[i][n - 1]) continue;
        return bin(i, 0, n);
    }
    return false;
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        def bin(row, i, j):
            mid = (i + j) // 2
            if target == matrix[row][mid]:
                return True
            if mid == i or mid == j:
                return False
            if matrix[row][mid] < target:
                return bin(row, mid, j)
            else:
                return bin(row, i, mid)
        for i in range(m):
            if target < matrix[i][0]:
                return False
            if target > matrix[i][-1]:
                continue
            return bin(i, 0, n)
        return False